lib/IRGen/GenConstant.cpp

```diff
@@ -0,0 +1,131 @@
+//===--- GenConstant.cpp - Swift IR Generation For Constant Values --------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements IR generation for constant values.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/IR/Function.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/IntrinsicInst.h"
+#include "llvm/IR/InlineAsm.h"
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallBitVector.h"
+#include "llvm/ADT/TinyPtrVector.h"
+
+#include "llvm/IR/Constant.h"
+#include "IRGenModule.h"
+#include "GenStruct.h"
+#include "TypeInfo.h"
+
+using namespace swift;
+using namespace irgen;
+
+/// Construct a ConstantInt from an IntegerLiteralInst.
+llvm::Constant *IRGenModule::emitConstantInt(swift::IntegerLiteralInst *ILI) {
+  APInt value = ILI->getValue();
+  BuiltinIntegerWidth width
+    = ILI->getType().castTo<BuiltinIntegerType>()->getWidth();
+
+  // The value may need truncation if its type had an abstract size.
+  if (width.isFixedWidth()) {
+    // nothing to do
+  } else if (width.isPointerWidth()) {
+    unsigned pointerWidth = getPointerSize().getValueInBits();
+    assert(pointerWidth <= value.getBitWidth()
+           && "lost precision at AST/SIL level?!");
+    if (pointerWidth < value.getBitWidth())
+      value = value.trunc(pointerWidth);
+  } else {
+    llvm_unreachable("impossible width value");
+  }
+
+  return llvm::ConstantInt::get(LLVMContext, value);
```

Comment:
This whole section can be much cleaner using early exits. In general in LLVM style, we like early exits b/c of the clarity value of control flow. For instance, compare what you have here with this:

```
if (width.isFixedWidth()) {
   return llvm::ConstantInt::get(LLVMContext, value);
}

if (!width.isPointerWidth()) {
   llvm_unreachable("impossible width value");
}

unsigned pointerWidth = getPointerSize().getValueInBits();
assert(pointerWidth <= value.getBitWidth()
           && "lost precision at AST/SIL level?!");
if (pointerWidth < value.getBitWidth())
     value = value.trunc(pointerWidth);
return llvm::ConstantInt::get(LLVMContext, value);
```